File: LetsPlant/app/src/main/java/com/e/letsplant/fragments/UsersFragment.java
package com.e.letsplant.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.MenuItemCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SearchView;
import android.widget.TextView;

import com.e.letsplant.App;
import com.e.letsplant.R;
import com.e.letsplant.adapters.UserAdapter;
import com.e.letsplant.data.User;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class UsersFragment extends MainFragment {
    private static Fragment instance = null;
    private List<User> mUsers;

    private TextView noUsersMessageTextView;
    private RecyclerView recyclerView;
    private UserAdapter userAdapter;

    private String userUid;
    private String userId;

    FirebaseAuth auth = App.getFirebaseAuthReference();
    FirebaseAuth.AuthStateListener authListener = new FirebaseAuth.AuthStateListener() {
        @Override
        public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
            FirebaseUser firebaseUser = firebaseAuth.getCurrentUser();
            if (firebaseUser != null) {
                userId = firebaseUser.getUid();
                String userEmail = firebaseUser.getEmail();
            }
        }
    };

    public UsersFragment() {
    }

    public static Fragment getInstance() {
        if (instance == null)
            instance = new UsersFragment();
        return instance;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        Objects.requireNonNull(((AppCompatActivity) requireActivity()).getSupportActionBar()).setTitle("");
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (firebaseAuth.getCurrentUser() != null)
            userUid = firebaseAuth.getCurrentUser().getUid();
        rootView = inflater.inflate(R.layout.fragment_users, container, false);

        noUsersMessageTextView = rootView.findViewById(R.id.noFriendsMessageTextView);

        recyclerView = rootView.findViewById(R.id.friendsRecyclerView);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        mUsers = new ArrayList<>();
        userAdapter = new UserAdapter(getContext(), mUsers);
        recyclerView.setAdapter(userAdapter);

        getAllUsers();

        return rootView;
    }

    @Override
    public void onCreateOptionsMenu(@NotNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.top_feed_bar_menu, menu);
        MenuItem item = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                if (!TextUtils.isEmpty(s.trim()) && !s.equals("")) {
                    searchUsers(s);
                } else {
                    getAllUsers();
                }
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                if (!TextUtils.isEmpty(s.trim()) && !s.equals("")) {
                    searchUsers(s);
                } else {
                    getAllUsers();
                }
                return false;
            }
        });

        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_search) {

            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void searchUsers(String query) {
        databaseReference.child(DB_USERS).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                mUsers.clear();
                for (DataSnapshot dataSnapshot : snapshot.getChildren()) {
                    User user = dataSnapshot.getValue(User.class);
                    if (!user.getId().equals(userUid)) {
                        if (user.getUsername().toLowerCase().contains(query.toLowerCase())) {
                            mUsers.add(user);
                        }
                    }
                }
                userAdapter.updateDataSet(mUsers);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void getAllUsers() {
        databaseReference.child(DB_USERS).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                mUsers.clear();
                for (DataSnapshot dataSnapshot : snapshot.getChildren()) {
                    User user = dataSnapshot.getValue(User.class);
                    if (firebaseAuth.getCurrentUser() != null) {
                        if (!user.getId().equals(userUid)) {
                            mUsers.add(user);
                        }
                    }
                }

                userAdapter.updateDataSet(mUsers);

                if (mUsers.size() == 0) {
                    noUsersMessageTextView.setVisibility(View.VISIBLE);
                } else {
                    noUsersMessageTextView.setVisibility(View.GONE);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }
}

File: LetsPlant/app/src/main/java/com/e/letsplant/fragments/UsersFragment.kt
package com.e.letsplant.fragments

import android.os.Bundle
import android.text.TextUtils
import android.view.*
import android.widget.SearchView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuItemCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.e.letsplant.R
import com.e.letsplant.adapters.UserAdapter
import com.e.letsplant.data.User
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import java.util.*

class UsersFragment : MainFragment() {
    private var mUsers: MutableList<User?>? = null
    private var noUsersMessageTextView: TextView? = null
    private var recyclerView: RecyclerView? = null
    private var userAdapter: UserAdapter? = null
    private var userUid: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        Objects.requireNonNull((requireActivity() as AppCompatActivity).supportActionBar)?.title = ""
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        if (firebaseAuth!!.currentUser != null) userUid =
            firebaseAuth!!.currentUser!!.uid
        rootView = inflater.inflate(R.layout.fragment_users, container, false)
        noUsersMessageTextView = rootView?.findViewById(R.id.noFriendsMessageTextView)
        recyclerView = rootView?.findViewById(R.id.friendsRecyclerView)
        recyclerView?.setHasFixedSize(true)
        recyclerView?.layoutManager = LinearLayoutManager(context)
        mUsers = ArrayList()
        userAdapter = UserAdapter(context, mUsers)
        recyclerView?.adapter = userAdapter
        allUsers
        return rootView
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.top_feed_bar_menu, menu)
        val item: MenuItem = menu.findItem(R.id.action_search)
        val searchView: SearchView = MenuItemCompat.getActionView(item) as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                if (!TextUtils.isEmpty(s.trim { it <= ' ' }) && s != "") {
                    searchUsers(s)
                } else {
                    allUsers
                }
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                if (!TextUtils.isEmpty(s.trim { it <= ' ' }) && s != "") {
                    searchUsers(s)
                } else {
                    allUsers
                }
                return false
            }
        })
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_search) {
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun searchUsers(query: String) {
        databaseReference!!.child(DB_USERS)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    mUsers!!.clear()
                    for (dataSnapshot: DataSnapshot in snapshot.children) {
                        val user: User? = dataSnapshot.getValue(
                            User::class.java
                        )
                        if (user?.id != userUid) {
                            if (user?.username?.lowercase(Locale.getDefault())?.contains(
                                    query.lowercase(
                                        Locale.getDefault()
                                    )
                                )!!
                            ) {
                                mUsers!!.add(user)
                            }
                        }
                    }
                    userAdapter!!.updateDataSet(mUsers)
                }

                override fun onCancelled(error: DatabaseError) {}
            })
    }

    private val allUsers: Unit
        get() {
            databaseReference!!.child(DB_USERS)
                .addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        mUsers!!.clear()
                        for (dataSnapshot: DataSnapshot in snapshot.children) {
                            val user: User? = dataSnapshot.getValue(
                                User::class.java
                            )
                            if (firebaseAuth!!.currentUser != null) {
                                if (user?.id != userUid) {
                                    mUsers!!.add(user)
                                }
                            }
                        }
                        userAdapter!!.updateDataSet(mUsers)
                        if (mUsers!!.size == 0) {
                            noUsersMessageTextView!!.visibility = View.VISIBLE
                        } else {
                            noUsersMessageTextView!!.visibility = View.GONE
                        }
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
        }

    companion object {
        var instance: Fragment? = null
            get() {
                if (field == null) field = UsersFragment()
                return field
            }
            private set
    }
}